cmake/configure_lvi_mitigation_build.cmake

```diff
@@ -1,25 +1,92 @@
 # Copyright (c) Open Enclave SDK contributors.
 # Licensed under the MIT License.
 
+macro (detect_compiler BINDIR CC)
+  set(GCC gcc)
+  set(CLANG clang)
+  if (${CC} STREQUAL CXX)
+    set(GCC g++)
+    set(CLANG clang++)
+  endif ()
+
+  # Set the COMPILER as the variable name (either C_COMPILER or CXX_COMPILER).
+  set(COMPILER "${CC}_COMPILER")
+
+  # Use the second level of unwrapping on COMPILER to get the compiler name.
+  # If the default compiler is found, return.
+  if (EXISTS "${BINDIR}/${${COMPILER}}")
+    return()
+  endif ()
+
+  if (NOT OE_IN_PACKAGE)
+    # Build OE. Fallback to gcc/g++.
+    set(${COMPILER} ${GCC})
+  else ()
+    # Build enclave applications. Try to search newer versions of clang/clang++.
+    # Be consistent to the logic implemented by samples/config.mk.
+    foreach (VERSION 9 8 7)
+      set(CLANG_VERSION "")
+      if (EXISTS "${BINDIR}/${${COMPILER}}-${VERSION}")
+        set(CLANG_VERSION ${VERSION})
+        break()
+      endif ()
+    endforeach ()
+    # Set the compiler if a version of clang/clang++ is found.
+    if (CLANG_VERSION)
+      set(${COMPILER} ${CLANG}-${CLANG_VERSION})
+    else ()
+      set(${COMPILER} ${GCC})
+    endif ()
+  endif ()
+endmacro ()
+
 # Adopt customized compiler wrappers for LVI mitigation.
-function (configure_lvi_mitigation_build LVI_MITIGATION_BINDIR)
-  if (NOT EXISTS "${LVI_MITIGATION_BINDIR}")
-    message(FATAL_ERROR "${LVI_MITIGATION_BINDIR} does not exist.")
+function (configure_lvi_mitigation_build)
+  cmake_parse_arguments(OE "IN_PACKAGE" "BINDIR" "" ${ARGN})
+
+  if (NOT EXISTS "${OE_BINDIR}")
+    message(FATAL_ERROR "${OE_BINDIR} does not exist.")
   endif ()
-  # Perfer clang over gcc.
-  if (EXISTS "${LVI_MITIGATION_BINDIR}/clang-7")
-    set(CMAKE_C_COMPILER
-        ${LVI_MITIGATION_BINDIR}/clang-7
-        PARENT_SCOPE)
-    set(CMAKE_CXX_COMPILER
-        ${LVI_MITIGATION_BINDIR}/clang++-7
-        PARENT_SCOPE)
+
+  if (NOT OE_IN_PACKAGE)
+    # Default to clang-7 when building SDK.
+    set(C_COMPILER clang-7)
+    set(CXX_COMPILER clang++-7)
+  else ()
+    # Default to clang when building enclave applications.
+    set(C_COMPILER clang)
```

Comment:
If clang was installed via apt and if a newer version of clang was also installed, then defaulting to clang would cause the older version (via apt) to be picked up. Can you confirm?

Reply:
Yes, I can confirm that. And I think this is an expected behavior. If user does not specify a version of clang via `CC`, The version of clang associated to the `clang` binary in the path will be invoked.

Reply:
@mingweishih I notice that on my Ubuntu there are binaries clang++-7 and clang-7 from clang 7, but no clang. Do we have "apt install clang" before this part?

Reply:
This is intentional that makes the installed clang as a default compiler. If clang is not installed, the following logic will try to use clang-9/8/7.